### PR TITLE
Fix #24959: Add Component Store for Template Builder

### LINKED ISSUE #24959: Template Builder: Implement component-store for state manegement
### Parent Issue

#24535 

### User Story 

As a user, I want to be able to add state management for the template builder component, so I can create and manage complex template configurations.

### Acceptance Criteria

- Implement the @ngrx/component-store library
- Move most of the logic to the service layer
- Ensure that the component is simple and easy to use

### Proposed Objective

Core Feature

### Proposed Priority

Priority 3 - Average

### External Links

N/A

### Assumptions & Initiation Needs

- The @ngrx/component-store library is already installed and configured

### Quality Assurance Notes & Workarounds

N/A

### Sub-Tasks & Estimates

- Implement @ngrx/component-store library
- Move logic to service layer
- Ensure component is simple and easy to use

### PR DESCRIPTION
**Pull request by zJaaal**
### Summary
📄⚙️🔧

This pull request introduces a new service, `DotTemplateBuilderStore`, that uses the `@ngrx/component-store` library to manage the state of the template builder component. It also refactors the `TemplateBuilderComponent` class to use this service and extracts the grid options and types to separate files for better readability and reusability.

> _Sing, O Muse, of the skillful coder who devised_
> _A store of state for the template builder, and thus_
> _Made easy the work of arranging rows and columns_
> _With the help of GridStack, the mighty grid-maker._

### Walkthrough
*  Refactor the template builder component to use the `@ngrx/component-store` library for state management ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L7-R9), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0R16), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L22-R27), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L178-R34), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L195-R50), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L210-R93), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L276-R119), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L366-R152), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L416-R206), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-f5608bc0e0cfd9dc606805e064c3a3de96ed347a844c8ef22f9ef831a0b80522R4), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-f5608bc0e0cfd9dc606805e064c3a3de96ed347a844c8ef22f9ef831a0b80522R10))
  * Add a new file `template-builder.store.ts` that defines the `DotTemplateBuilderStore` class, which extends the `ComponentStore` class and provides selectors and updaters for the items array ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-a32a0fcbcf1b026389eade1b0cd88e8838bbfacb92905fe9da2bea7f72b7c9abR1-R179))
  * Inject the `DotTemplateBuilderStore` service into the `TemplateBuilderComponent` class and use its methods to update and select the state ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L195-R50))
  * Remove the unused import of `GridStackOptions` and the logic for handling actions and state changes using a `Subject` and a `scan` operator ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L7-R9), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L22-R27), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L178-R34))
  * Replace the `GridStackWidget` type with the `DotGridStackWidget` type, which has more properties and extends the former ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L22-R27))
  * Call the `init` method of the store service with the starter array of items, and assign the `items$` observable to the `items$` selector of the store service ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L195-R50))
  * Update the logic for adding, moving, and updating rows and columns to call the store service methods, such as `addRow`, `addColumn`, `moveRow`, and `updateColumn` ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L210-R93), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L276-R119), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L366-R152))
  * Add the `OnDestroy` lifecycle hook and the `ngOnDestroy` method to the component class, which calls the `destroy` method of the grid instance with the `true` argument ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0R16), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L416-R206))
  * Add the `DotTemplateBuilderStore` service to the `providers` array of the `TemplateBuilderModule` class, which registers the service as a dependency ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-f5608bc0e0cfd9dc606805e064c3a3de96ed347a844c8ef22f9ef831a0b80522R10))
* Extract the grid and subgrid options and the helper functions to a separate file `options.ts`, which is imported by the component class ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L210-R93), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L276-R119), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L366-R152), [link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-8fbd35880751ae16a8b566cb1a569aebb4c79c51265e892b9fa1273b38896abbR1-R38))
  * Define the constants for the `gridOptions` and `subGridOptions`, which are used by the `GridStack` library to initialize and configure the grids ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-8fbd35880751ae16a8b566cb1a569aebb4c79c51265e892b9fa1273b38896abbR1-R38))
  * Define the helper functions `isAColumnWidget` and `isARowWidget`, which are used by the `acceptWidgets` option to determine which widgets can be dropped on which grids ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-8fbd35880751ae16a8b566cb1a569aebb4c79c51265e892b9fa1273b38896abbR1-R38))
* Add a new file `types.ts` that defines the interfaces and types for the template builder component and store, which extend or modify the existing `GridStack` types with additional properties and features ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-6f6b63bcffd3a8cc2df185547ac56a2f01f9f33dfca7a536b29c1e4b832cfaf0R1-R21))
  * Define the `DotGridStackOptions`, `DotContainers`, `DotGridStackWidget`, and `DotTemplateBuilderState` interfaces ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-6f6b63bcffd3a8cc2df185547ac56a2f01f9f33dfca7a536b29c1e4b832cfaf0R1-R21))
* Remove a commented out console log statement, which is not useful for debugging or documentation purposes ([link](https://github.com/dotCMS/core/pull/24983/files?diff=unified&w=0#diff-d12c3fc3a47505ab4b846cfc5090445b7597e2acd4c10382ff0dc22f8c0d9ff0L351))

### Screenshot

https://github.com/dotCMS/core/assets/63567962/22f47f76-82dc-4de7-8fdf-221d452dca66

### CODE REVIEW

File: core-web/libs/template-builder/src/lib/components/template-builder/store/template-builder.store.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            if (addNodeRowIndex > -1) {
+                if (itemsCopy[addNodeRowIndex].subGridOpts)
+                    itemsCopy[addNodeRowIndex].subGridOpts.children.push(nodeToAdd); // Add the node
+                else itemsCopy[addNodeRowIndex].subGridOpts = { children: [nodeToAdd] };
+            }
+
+            return {
+                items: itemsCopy
+            };
+        }
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/store/template-builder.store.ts:128]: We can use the `Utils.xxx` from gridstack for this complex transformations.

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.ts [hunk 2]
```diff
@@ -207,132 +61,62 @@ export class TemplateBuilderComponent implements OnInit, AfterViewInit {
...
-                actionType: 'yMove',
-                payload: nodes.map((node) => ({
+        this.grid = GridStack.init(gridOptions).on('change', (_: Event, nodes: GridStackNode[]) => {
+            this.store.moveRow(
+                nodes.map((node) => ({
                     y: node.y,
                     id: node.id as string,
                     parentId: node.grid?.parentGridItem?.id as string
                 }))
-            };
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.ts:257]: Let me store take care of this... lets keep the business logic in one place.

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.ts [hunk 3]
```diff
@@ -413,7 +162,46 @@ export class TemplateBuilderComponent implements OnInit, AfterViewInit {
...
+                parentId: newNode.grid?.parentGridItem?.id as string,
+                w: newNode.w,
+                h: newNode.h,
+                x: newNode.x,
+                y: newNode.y,
+                id: String(ids++)
+            });
+
+            newNode.grid?.removeWidget(newNode.el as GridStackElement, true);
+        }
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.ts:205]: Let's leave the store take care of this, we send the whole thing and the store (or some external util) take care of this. Let's keep the component as dumb as possible.

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/options.ts
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/options.ts:1]: `gridtstack-options.ts`

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/types.ts
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/types.ts:1]: `/template-builder/models/models.ts`

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/store/template-builder.store.ts (hunk 1 above)
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/store/template-builder.store.ts]: made a whole refactor of this code to increase readability

**Review by rjvelazco — APPROVED** [review]
.

**Review by rjvelazco — CHANGES_REQUESTED** [review]
.

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.ts [hunk 4]
```diff
@@ -192,148 +44,57 @@ export class TemplateBuilderComponent implements OnInit, AfterViewInit {
...
+                    children: [{ x: 0, y: 0, w: 4, id: uuid() }]
                 }
             }
         ];
 
-        // HERE ARE CHANGES
-        this.items$ = this.mySub$.pipe(
-            scan((acc: GridStackWidget[], { widgetType, actionType, payload }: Action) => {
-                const newAcc = structuredClone(acc);
+        this.store.init(starter);
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.ts]: We still need another service to parse the backend data, so I'm initializing the state here for testing purposes

File: core-web/libs/template-builder/src/lib/components/template-builder/store/template-builder.store.ts
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/store/template-builder.store.ts:1]: Will remain as it is but with util function to improve readability as we discussed

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import { GridStackOptions } from 'gridstack';
+
+/**
+ * Check if the element is a column widget by checking the data-widget-type attribute
+ *
+ * @param {Element} el
+ * @return {*}  {boolean}
+ */
+function isAColumnWidget(el: Element): boolean {
+    return el.getAttribute('data-widget-type') === 'col' || el.classList.contains('sub');
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-options.ts:10]: I like the idea, I'll add it

**Review by manuelrojas — APPROVED** [review]

**Pull request merged**